packages/medusa/src/services/fulfillment-provider.ts

```diff
@@ -0,0 +1,177 @@
+import { MedusaError } from "medusa-core-utils"
+import BaseFulfillmentService from "medusa-interfaces/dist/fulfillment-service"
+import { EntityManager } from "typeorm"
+import { TransactionBaseService } from "../interfaces"
+import {
+  Cart,
+  Fulfillment,
+  FulfillmentProvider,
+  LineItem,
+  Order,
+  Return,
+  ShippingMethod,
+  ShippingOption,
+} from "../models"
+import { FulfillmentProviderRepository } from "../repositories/fulfillment-provider"
+import { CreateFulfillmentOrder } from "../types/fulfillment"
+import { FulfillmentOptions } from "../types/fulfillment-provider"
+import { MedusaContainer } from "../types/global"
+
+type FulfillmentProviderContainer = MedusaContainer & {
+  fulfillmentProviderRepository: typeof FulfillmentProviderRepository
+  manager: EntityManager
+}
+/**
+ * Helps retrive fulfillment providers
+ */
+class FulfillmentProviderService extends TransactionBaseService<FulfillmentProviderService> {
+  protected manager_: EntityManager
+  protected transactionManager_: EntityManager | undefined
+  protected readonly container_: FulfillmentProviderContainer
+
+  constructor(container: FulfillmentProviderContainer) {
+    super(container)
+
+    const { manager } = container
+
+    this.container_ = container
+    this.manager_ = manager
+  }
+
+  async registerInstalledProviders(providers: string[]): Promise<void> {
+    return await this.atomicPhase_(async (manager) => {
+      const { fulfillmentProviderRepository } = this.container_
+      const fulfillmentProviderRepo = manager.getCustomRepository(
+        fulfillmentProviderRepository
+      )
+      await fulfillmentProviderRepo.update({}, { is_installed: false })
+
+      for (const p of providers) {
+        const n = fulfillmentProviderRepo.create({ id: p, is_installed: true })
+        await fulfillmentProviderRepo.save(n)
+      }
+    })
+  }
+
+  async list(): Promise<FulfillmentProvider[]> {
+    const { manager, fulfillmentProviderRepository } = this.container_
+    const fpRepo = manager.getCustomRepository(fulfillmentProviderRepository)
+
+    return await fpRepo.find({})
+  }
+
+  async listFulfillmentOptions(
+    provider_ids: string[]
+  ): Promise<FulfillmentOptions[]> {
+    return await Promise.all(
+      provider_ids.map(async (p) => {
+        const provider = await this.retrieveProvider(p)
+        return {
+          provider_id: p,
+          options:
+            (await provider.getFulfillmentOptions()) as unknown as Record<
+              string,
+              unknown
+            >[],
+        }
+      })
+    )
+  }
+
+  /**
+   * @param {string} provider_id - the provider id
+   * @return {BaseFulfillmentService} the payment fulfillment provider
+   */
+  retrieveProvider(provider_id: string): BaseFulfillmentService {
+    try {
+      return this.container_[`fp_${provider_id}`]
```

Comment:
todo: this should be the type of the container to restrict the access. You can look here https://github.com/medusajs/medusa/blob/master/packages/medusa/src/services/notification.ts#L20